BUG: pd.Index with pd.array does not respect dtype.

### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.


### Reproducible Example

```
python
import pandas as pd

# Buggy behavior
print(pd.Index(pd.array([1,0]), dtype='object'))
# Index([1, 0], dtype='Int64')

# Correct behavior
print(pd.Index([1,0], dtype='object'))
# Index([1, 0], dtype='object')
```


### Issue Description

This issue affects the latest version: '1.4.0.dev0+1574.g46ddb8ef88'

In older versions, such as 1.3.5, have the first argument as a pandas array does not affect the output dtype. With the latest version, this is not the case. The output dtype should be the dtype requested in the argument, in this case 'object'.

### Expected Behavior

import pandas as pd

# Buggy behavior
```
print(pd.Index(pd.array([1,0]), dtype='object'))
# Index([1, 0], dtype='object')
```

### Installed Versions

<details>

>>> pd.show_versions()
/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/_distutils_hack/__init__.py:24: UserWarning: Distutils was imported before Setuptools, but importing Setuptools also replaces the `distutils` module in `sys.modules`. This may lead to undesirable behaviors or errors. To avoid these issues, avoid using distutils directly, ensure that setuptools is installed in the traditional way (e.g. not an editable install), and/or make sure that setuptools is always imported before distutils.
  warnings.warn(
/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/_distutils_hack/__init__.py:36: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/pandas/util/_print_versions.py", line 109, in show_versions
    deps = _get_dependency_info()
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/pandas/util/_print_versions.py", line 88, in _get_dependency_info
    mod = import_optional_dependency(modname, errors="ignore")
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/pandas/compat/_optional.py", line 116, in import_optional_dependency
    module = importlib.import_module(name)
  File "/opt/miniconda3/envs/nightly/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/setuptools/__init__.py", line 8, in <module>
    import _distutils_hack.override  # noqa: F401
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/_distutils_hack/override.py", line 1, in <module>
    __import__('_distutils_hack').do_override()
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 73, in do_override
    ensure_local_distutils()
  File "/opt/miniconda3/envs/nightly/lib/python3.10/site-packages/_distutils_hack/__init__.py", line 61, in ensure_local_distutils
    assert '_distutils' in core.__file__, core.__file__
AssertionError: /opt/miniconda3/envs/nightly/lib/python3.10/distutils/core.py

</details>

--- a/pandas/core/indexes/base.py
+++ b/pandas/core/indexes/base.py
@@ -453,6 +453,9 @@ class Index(IndexOpsMixin, PandasObject):
                 if dtype is not None:
                     return result.astype(dtype, copy=False)
                 return result
+            elif dtype is not None:
+                # GH#45206
+                data = data.astype(dtype, copy=False)
 
             disallow_kwargs(kwargs)
             data = extract_array(data, extract_numpy=True)

--- a/pandas/tests/indexes/common.py
+++ b/pandas/tests/indexes/common.py
@@ -422,9 +422,11 @@ class Base:
             #  fails for IntervalIndex
             return
 
+        is_ea_idx = type(index) is Index and not isinstance(index.dtype, np.dtype)
+
         assert index.equals(index)
         assert index.equals(index.copy())
-        if not (type(index) is Index and not isinstance(index.dtype, np.dtype)):
+        if not is_ea_idx:
             # doesn't hold for e.g. IntegerDtype
             assert index.equals(index.astype(object))
 
@@ -432,7 +434,7 @@ class Base:
         assert not index.equals(np.array(index))
 
         # Cannot pass in non-int64 dtype to RangeIndex
-        if not isinstance(index, RangeIndex):
+        if not isinstance(index, RangeIndex) and not is_ea_idx:
             same_values = Index(index, dtype=object)
             assert index.equals(same_values)
             assert same_values.equals(index)

--- a/pandas/tests/indexes/test_index_new.py
+++ b/pandas/tests/indexes/test_index_new.py
@@ -25,6 +25,7 @@ from pandas import (
     Series,
     TimedeltaIndex,
     Timestamp,
+    array,
     date_range,
     period_range,
     timedelta_range,
@@ -158,6 +159,13 @@ class TestIndexConstructorInference:
 
 class TestDtypeEnforced:
     # check we don't silently ignore the dtype keyword
+
+    def test_constructor_object_dtype_with_ea_data(self, any_numeric_ea_dtype):
+        # GH#45206
+        arr = array([0], dtype=any_numeric_ea_dtype)
+
+        idx = Index(arr, dtype=object)
+        assert idx.dtype == object
 
     @pytest.mark.parametrize("dtype", [object, "float64", "uint64", "category"])
     def test_constructor_range_values_mismatched_dtype(self, dtype):

